pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -199,13 +200,6 @@ func (jm *ControllerV2) sync(ctx context.Context, cronJobKey string) (*time.Dura
 	cronJobCopy, requeueAfter, updateStatus, err := jm.syncCronJob(ctx, cronJob, jobsToBeReconciled)
 	if err != nil {
 		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KObj(cronJob), "err", err)
-		if updateStatus {
```

Comment:
without much experience on this controller this change looks like an important behavior change in the controller

Reply:
There is exactly the same if-statement below, which is going to be executed unless there is a panic in the code, as there are no return-statements in-between: https://github.com/kubernetes/kubernetes/blob/16173c23baa3a6a7ad19233c963d165dc7f602d9/pkg/controller/cronjob/cronjob_controllerv2.go#L208-L214

Reply:
Could we nevertheless make attempt to update the status here, but with different error handling, and a comment in both places about the situation?

Logging that the status update failed is helpful; it's just the `return` statements that create a problem.

Any other logic refactor that ensures we attempt `cleanupFinishedJobs()` is also fine by me. Ideally we don't drop any observability along with that change.

Reply:
Maybe we should set `updateStatus` to true, but defer the update?

I'd like to see a clear comment that shows why this change is right. It might be, but I want the code and PR to explain why.

Reply:
We can still remove the copy-paste like this: https://github.com/ASverdlov/kubernetes/commit/675e411ae6958bc6b07e7ae104480e316e403fd4.

Attempt on a clear comment on why it would work follows: let's compare 2 versions: because the following 2 versions of code are equivalent:
```go
	cronJobCopy, requeueAfter, updateStatus, err := jm.syncCronJob(ctx, cronJob, jobsToBeReconciled)
	if err != nil {
		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KObj(cronJob), "err", err)

                COPY_PASTE()
		return nil, err
	}

        COPY_PASTE()

	if requeueAfter != nil {
		logger.V(4).Info("Re-queuing cronjob", "cronjob", klog.KObj(cronJob), "requeueAfter", requeueAfter)
		return requeueAfter, nil
	}
	// this marks the key done, currently only happens when the cronjob is suspended or spec has invalid schedule format
	return nil, nil

```
and

```go
	cronJobCopy, requeueAfter, updateStatus, reconciliationErr := jm.syncCronJob(ctx, cronJob, jobsToBeReconciled)
	if err != nil {
		logger.V(2).Info("Error reconciling cronjob", "cronjob", klog.KObj(cronJob), "err", err)
	}
        
        COPY_PASTE()

	if requeueAfter != nil {
		logger.V(4).Info("Re-queuing cronjob", "cronjob", klog.KObj(cronJob), "requeueAfter", requeueAfter)
		return requeueAfter, reconciliationErr
	}
	// this marks the key done, currently only happens when the cronjob is suspended or spec has invalid schedule format
	return nil, reconciliationErr

```

* In case `jm.syncCronJob` doesn't return an error, they are line-by-line the same (if you ignore renaming of the error variable.
* In case `jm.syncCronJob` does return an error, the only difference is: latter code also returns a `requeueAfter` in case it was set. But `jm.syncCronJob()` never returns a `requeueAfter != nil` when it returns an error, so in practice it doesn't change the current behaviour.